Add predicate syntax

dario23 opened the pull request.

fpoli commented on prusti-specs/src/lib.rs:
```diff
@@ -421,3 +421,37 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
+pub fn predicate(attr: TokenStream, tokens: TokenStream) -> TokenStream {
+    if !attr.is_empty() {
+        // TODO: what here? error? re-emit the previous attributes? not sure
+        //       what the semantics for that are supposed to be..
```
Do you know an example that triggers this? In the other macros we just call the first parameter `_attr` and we never use it.

dario23 replied:
turns out i misunderstood what this parameter is used for; just ignoring it seems like the best idea, given that's how the rest of the codebase deals with this, will change.

fpoli replied:
`#[predicate(123)]` looks like an example (see [here](https://doc.rust-lang.org/reference/procedural-macros.html#attribute-macros)), so raising a compilation error as you mention in the TODO sounds good. The `pure` and `trusted` macros don't do this check, but I think they should.

fpoli replied:
To return a compilation error you can use `return syn::Error::new(span, "error message").to_compile_error();`. I hope you can obtain the span from `attr`, somehow.

dario23 replied:
i'd be happy to address this, but would open a separate PR for that, what do you think?

fpoli replied:
For `pure` and `trusted` yes.

dario23 replied:
https://github.com/viperproject/prusti-dev/pull/391

fpoli commented on prusti-specs/src/lib.rs:
```diff
@@ -421,3 +421,37 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
+                #[requires(#assertion_tokens)]
+                #[pure]
+                // TODO: is `true` the only possibility here? do we require
+                //       all predicates to be `-> bool`?
```
Is the comment is ignored by `parse_quote_spanned`? If not or if unsure, place it before the macro call. Anyway, I think it's correct to only have boolean predicates.

Aurel300 replied:
Comments are not part of the AST (`syn`'s or Rust's), so these are definitely stripped during lexing of `lib.rs`.

fpoli commented on prusti-tests/tests/parse/ui/predicate_fail-1.rs:
```diff
@@ -0,0 +1,14 @@
+// compile-flags: -Pprint_desugared_specs=true -Pprint_typeckd_specs=true -Pno_verify=true -Phide_uuids=true
+// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
+// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"
```
~~These two lines should not be needed (here an in other files).~~ I found the `.stdout`

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -421,3 +421,37 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
+                group.stream()
+            } else {
+                unreachable!("A function's block must be a brace-delimited `TokenTree::Group`")
+            };
```
Why match like this? Is this to strip the braces? I wonder if you can't simply do:
```suggestion
            let body = pred_func.block;
```
And then `#[requires #body ]`?

dario23 replied:
ah, very nice way to get rid of the braces, yes. didn't think of that before.

Aurel300 commented on prusti-tests/tests/verify/fail/predicates/doesnt-hold.rs:
```diff
@@ -0,0 +1,11 @@
+    false
+}
+
+#[ensures(false_p())] //~ ERROR precondition of pure function call might not hold
```
It's good that this raises an error and with a reasonable position, but now that I think about it, the error message is actually a bit confusing. "precondition of pure function" here refers to `false_p`, which is specified without a precondition as far as the user is concerned. Changing this would actually be a bit more involved:

- during macro processing of `#[predicate]`, mark the resulting function with some new prusti-reserved attribute
- after typing, predicates should be identified by the `SpecCollector`
  - `ProcedureSpecifications` should get a new field `predicate: bool` if the attribute was found in the HIR
- `ErrorCtxt` needs a new variant for predicate calls
- `encode_specified_pure_function_call` (in `procedure_encoder`) needs to identify when a pure function call is a call to a predicate (which will now be stored in the specification set)

We can wait until Monday if this is unclear, it also makes me a bit uneasy to introduce a relatively large set of changes to fix one error message…

dario23 replied:
yep, i think i understand the problems you see here, but i agree with the "large set of changes to fix one error message". i'd rather get this feature landed at all first, and do more involved stuff like this subsequently.

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -118,12 +119,22 @@ fn generate_for_requires(attr: TokenStream, item: &untyped::AnyFnItem) -> Genera
+
+/// Generate spec items and attributes to typecheck and later retrieve "ensures_result_eq" annotations.
+fn generate_for_ensures_result_eq(attr: TokenStream, item: &untyped::AnyFnItem) -> GeneratedResult {
+    let assertion = parse_quote_spanned!{item.span() => result == (#attr)};
```
This is not enough, as it still produces e.g. `result == forall(...)` and expects the preparser to be able to deal with it, which it cannot. The approach here is to parse the attribute as an *expression*, then add the `result == ...` over the parsed tree (without relying on the preparser).

dario23 replied:
and by *expression*, you mean a `prusti_specs::specifications:untyped::Assertion`, right?

Aurel300 replied:
Yeah, sorry, I was confused due to this being different in VIR. But this raises a good point, I'm not sure you can actually build an assertion that is `AssertionKind::Expr` representing `result == <another assertion>`. I'll need to think about this a bit more…

dario23 replied:
put up a WIP of reworking to handle predicates separately all the way through, still some work needed.

fpoli commented on prusti-specs/src/lib.rs:
```diff
@@ -435,3 +436,61 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
+            let assertion = handle_result!(rewriter.parse_assertion(spec_id, pred_tokens));
+
+            let item_name = syn::Ident::new(
+                &format!("prusti_{}_item_{}_{}", "pred", pred_func.sig.ident, spec_id),
```
"pred" can be inlined in the format string.

dario23 committed:
Predicate Syntax: Add desugaring and parse checks

This is (sort of) step 1/3 of implementing this. In this change, we have
implemented the desugaring of predicate functions into an empty function
and additional generated functions to typecheck the arguments of the
predicate.

The next steps will be the new `SpecChecker` and the actual encoding to
Viper pure functions. The SpecChecker checks that predicate functions
are never actually called and emits clear error messages for that case.
It could in the future have more tasks than just this.
The encoding to Viper will reuse large parts of what's already there for
`ProcedureSpecification`s, as those are encoded to pure functions
already and we'll duplicate a lot of code otherwise.

dario23 committed:
Add SpecChecker

For now the only check here is that predicate functions are never
called. Other checks may be added in the future.

This is step 2/3, as outlined in the previous commit. The final step
will be the translation of predicates to viper pure functions.

dario23 commented on prusti-specs/src/lib.rs:
```diff
@@ -435,3 +436,65 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
+            let mut rewriter = rewriter::AstRewriter::new();
+            let spec_id = rewriter.generate_spec_id();
+            let assertion = handle_result!(rewriter.parse_assertion(spec_id, pred_tokens));
+            let predicate = untyped::Predicate::new(assertion);
```
maybe we don't need this wrapper type, and should just go ahead with the inner assertion here?

dario23 replied:
removed it in 5e6a2469e

fpoli commented on prusti-interface/src/specs/checker.rs:
```diff
@@ -0,0 +1,118 @@
+pub struct SpecChecker<'tcx> {
+    tcx: TyCtxt<'tcx>,
+
+    predicates: HashMap<DefId, Span>,
```
Could you add a `///` comment for the `predicates` field and the `SpecChecker` type?

dario23 replied:
- [x] done in 34cd995b

dario23 commented on prusti-interface/src/specs/checker.rs:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    fn visit_expr(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        if let hir::ExprKind::Call(ref callee_expr, _) = ex.kind {
```
not just calls, even just refering to it should be disallowed (i.e. get rid of the outermost if, i think)

dario23 replied:
0bab6b18

dario23 commented on prusti-interface/src/specs/checker.rs:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    fn visit_trait_item(&mut self, _: &'tcx hir::TraitItem<'tcx>) {
+        // nothing here
```
do we need some default action here to visit sub-items? should also add a test for it.

dario23 replied:
seems the ItemLikeVisitor is too shallow here, only visiting toplevel items; rewriting it using `intravisit::Visitor`.

fpoli commented on prusti-interface/src/specs/checker.rs:
```diff
@@ -0,0 +1,118 @@
+        id: hir::HirId,
+    ) {
+        // Stop checking inside `prusti::spec_only` functions
+        if has_prusti_attr(fk.attrs(), "spec_only") {
```
We have `has_spec_only_attr(fk.attrs())`.

dario23 committed:
Remove Predicate wrapper type

it is not needed, we only care about the assertion inside, and had no
additional methods that would justify a separate type.

dario23 committed:
Add doc comments for SpecChecker and its fields

dario23 committed:
Check any reference to predicate functions, not just calls

dario23 committed:
Check nested items in SpecChecker

squash

dario23 committed:
Collect predicate specifications in SpecCollector

This also adds a ui-test for the parsing, desugaring and collecting.

Also we have to update a number of ui testcases to include the new
`predicate_body` field.

fpoli commented on prusti-interface/src/specs/checker.rs:
```diff
@@ -0,0 +1,160 @@
+    predicates: HashMap<DefId, Span>,
+
+    /// Span of call and definition of predicates illegally called, collected in the second pass.
+    pred_calls: Vec<(Span, Span)>,
```
`pred_usages` would be a better name, because of 0bab6b182dbfdf47bec7292b9a83566a0f20e468. Similarly, the comment could say "usage" instead of "call".

dario23 replied:
updated in d4e946de.

i switched to many separate commits, mainly because i hope it makes reviewing easier; at least it helps me keep an overview. if you want, we can switch back though, e.g. for nicer git-history or such.

dario23 committed:
Update variable names, docs for predicate usage checks

The checks are no longer just about (illegally) calling predicate
functions, but any usage of them, so update names and comments
accordingly.

dario23 committed:
Encoding of predicate functions

Encoded very similar to other pure functions, so most of the changes are
in pure_function_encoder, just the call is in the encoder.rs file

fpoli commented on prusti-viper/src/encoder/encoder.rs:
```diff
@@ -1282,7 +1282,12 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
+                    .run_if_err(cleanup)?;
+                pure_function_encoder.encode_predicate_function(predicate_body_encoded)
+                    .run_if_err(cleanup)?
+            } else if self.is_trusted(proc_def_id) {
```
This makes me wonder, what happens if the user specifies both `#[predicate]` and `#[trusted]`? Would it be easy to generate an error message e.g. in the `SpecChecker`?

fpoli commented on prusti-viper/src/encoder/encoder.rs:
```diff
@@ -1282,7 +1282,12 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
-            let function = if self.is_trusted(proc_def_id) {
+            let function = if let Some(predicate_body) = self.get_predicate_body(proc_def_id) {
+                let predicate_body_encoded = self.encode_assertion(predicate_body, procedure.get_mir(), None, &[], None, false, None, ErrorCtxt::GenericExpression)
+                    .run_if_err(cleanup)?;
```
*  I think that `targets_are_values` should be `true` and that you should provide `target_args` ([here](https://github.com/viperproject/prusti-dev/blob/e207038b94cb8e5499b309de7ec52765b6e2404b/prusti-viper/src/encoder/pure_function_encoder.rs#L289-L293) is an example of how to build them). See [here](https://github.com/viperproject/prusti-dev/blob/e207038b94cb8e5499b309de7ec52765b6e2404b/prusti-viper/src/encoder/spec_encoder.rs#L32-L61) for a description of the arguments.
* I would prefer to move ~~the `get_predicate_body` and~~ the construction of `predicate_body_encoded` into the `encode_predicate_function` function, to keep `encoder.rs` clean.
* This line is very long; consider placing one argument per line.

dario23 commented on prusti-interface/src/utils.rs:
```diff
@@ -270,7 +270,7 @@ pub fn has_prusti_attr(attrs: &[ast::Attribute], name: &str) -> bool {
         ast::AttrKind::Normal(ast::AttrItem {
                                   path: ast::Path { span: _, segments, tokens: _ },
-                                  args: ast::MacArgs::Empty,
+                                  args: _, // ast::MacArgs::Empty,
```
remove from PR

dario23 replied:
rebased onto latest master and undid this change.

dario23 replied:
un-removing this; the issue is that `has_prusti_attr` only checked for things like `prusti::spec_only` which don't have values, but i need it to also match `prusti::pred_spec_id_ref = $(UUID)`, so i think this should be `args: _` after all. i'll remove the commented-out previous code of course.

fpoli replied:
If you use `args: _` remember to remove this comment

https://github.com/viperproject/prusti-dev/blob/50d1941a4e6ab78ebab853b48987ccd9b629f880/prusti-interface/src/utils.rs#L268

dario23 committed:
Add verify-fail predicates test case

dario23 committed:
Keep matching any args in has_prusti_attr

We used to only match on attributes without values like
`#[prusti::spec_only]`, but need to match
`#[prusti::pred_spec_id_ref = $(UUID)]` now as well, so ignore the args
and just return whether the given attribute exists

dario23 committed:
Refactor `predicate` to include check for other attrs

This change includes parsing the function as an `untyped::AnyFnItem`
instead of `syn::Item`, some more checks beside the main one for other
prusti attributes and as a side effect improving the error messages to
only list the actual errors we care about, not syntactic problems when
parsing e.g. a static item as function.

dario23 committed:
Use generate_spec_item_fn in predicate

This de-duplicates some of the code

dario23 committed:
Update comment

dario23 committed:
Consistently put #[allow(..)] first in pred desugaring

This hopefully fixes the tests

Aurel300 commented on prusti-interface/src/specs/checker.rs:
```diff
@@ -0,0 +1,160 @@
+}
+
+// Second predicate checks visitor: check any references to predicate functions
+// from non-specification code
```
```suggestion
/// Second predicate checks visitor: check any references to predicate functions
/// from non-specification code
```

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -60,6 +60,17 @@ pub fn rewrite_prusti_attributes(
+    {
+        return syn::Error::new(
+            item.span(),
+            "`#[predicate]` is incompatible with other prusti attributes",
```
```suggestion
            "`#[predicate]` is incompatible with other Prusti attributes",
```

(Unsure if we are consistent about this.)

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -435,3 +450,75 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
+        syn::parse2(tokens)
+            .map_err(|e| syn::Error::new(
+                e.span(),
+                "you can only use `#[predicate]` on function definitions"
```
```suggestion
                "`#[predicate]` can only be used on function definitions"
```

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -435,3 +450,75 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
+    // check for other attributes -- #[predicate] is incompatible with all others currently
+    // make sure to also update the check in `rewrite_prusti_attributes`
+    let other_prusti_attrs = extract_prusti_attributes(&mut item);
+    if other_prusti_attrs.count() != 0 {
```
```suggestion
    if !other_prusti_attrs.is_empty() {
```

dario23 replied:
`extract_prusti_attributes` returns an `impl Iter`, not a `Vec`, which doesn't seem to have an `is_empty` method.

Aurel300 commented on prusti-specs/src/lib.rs:
```diff
@@ -435,3 +450,75 @@ pub fn extern_spec(_attr: TokenStream, tokens:TokenStream) -> TokenStream {
+    if other_prusti_attrs.count() != 0 {
+        return syn::Error::new(
+            item.span(),
+            "`#[predicate]` is incompatible with other prusti attributes",
```
```suggestion
            "`#[predicate]` is incompatible with other Prusti attributes",
```

dario23 committed:
Some spelling/minor fixes

dario23 committed:
Adjust tests

Also add tests for the incompatibility changes
